### ISSUE TITLE
trial's force-garbage-collection flag doesn't affect testSuite/test_suite generated tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2793]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2793|
|Type|defect|
|Created|2007-08-20 12:46:26Z|
|Branch|https://github.com/twisted/twisted/tree/force-gc-decorator-2793-2|

```
exarkun@charm:~$ trial --force-gc test_foo.py
Running 1 tests.
test_foo
  FooTests
    test_bar ...                                                           [OK]

-------------------------------------------------------------------------------
Ran 1 tests in 0.001s

PASSED (successes=1)
exarkun@charm:~$ cat test_foo.py
import gc
gc.collect = lambda: 1/0

from unittest import TestSuite, TestCase

class FooTests(TestCase):
    def test_bar(self):
        pass

def testSuite():
    return TestSuite([FooTests('test_bar')])
exarkun@charm:~$
```

Attachments:

* [force-gc-decorator-2793-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5a5/5a5f42d2a90c9103167e8ebb5dd94727134b1155/568284f24b9133c50beafde76ff0873fd934463d.diff) (10326 bytes) - added by jml on 2007-12-23 00:53:45Z - 
* [force-gc-review.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5a5/5a5f42d2a90c9103167e8ebb5dd94727134b1155/9074546329101bafe3d3c0b9ec964ae66bb7f43f.diff) (6928 bytes) - added by jml on 2007-12-24 00:25:46Z - Changes from review

<details><summary>Searchable metadata</summary>

```
trac-id__2793 2793
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_force_gc_decorator_2793_2 branches/force-gc-decorator-2793-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1187613986000000 1187613986000000
changetime__1199050228000000 1199050228000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

(In [21744]) Branching to 'force-gc-2793-2'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

Absolutely no branch. Waiting for pre-requisite unittest improvements.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

(In [22128]) Branching to 'force-gc-decorator-2793'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

(In [22130]) Branching to 'force-gc-decorator-2793-2'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml removed owner|
|-|-|

Fixed this by creating a new decorator for `TestCase` that does the force-gc stuff. The runner now applies that decorator at runtime.

Some tests have been deleted without providing equivalent replacements. I think the original tests were too fragile. Proper tests for the way that the script options affect the run behaviour should wait until we can make the script a little more sane.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

Attached diff generated from branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jml|
|-|-|

```_ForceGarbageCollectionDecorator``` calls ```_installObserver``` on ```self.original```.  ```self.original``` is only guaranteed (or likely) to provide ```ITestCase``` though.  So, ```_ForceGarbageCollectionDecorator``` will only work if it is applied directly to a ```TestCase```.  If there is another decorator in between, it will fail.  I think the logged error collection functionality should be factored into a separate class which can be shared between ```TestCase``` and ```_ForceGarbageCollectionDecorator```.

```TrialRunner``` doesn't document its new (_public_) attribute, ```forceGarbageCollection```.  I wonder if it wouldn't be better for ```TrialRunner``` to take a list of decorators to apply instead of a new flag, anyway.

I think there can probably still be something like the removed ```TestGarbageCollect``` test cases.  ```TrialRunner``` knows about _force-gc_; it shouldn't be very difficult to assert something about an instance of it as created through the trial script code.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

Agreed on first point.

`forceGarbageCollection` should not be public. (Aren't incidental undocumented, untested attributes assumed to be non-public?)

There are lots of designs for `TrialRunner` that would be better than the one we have now. I'd like to continue incrementally factoring the innards before I make a significant change to its API.

If you insist, I can add tests like those you say, except they will monkey patch `gc.collect` and try to run some code, since I do _not_ want to enshrine the implementation detail of the attribute with tests. More explicitly, I want to test the interface, no more. 

I would rather not add the test -- too much work for not enough gain. Actually, I don't want to optimize for work, I want to optimize for velocity, so I'll do it. But please interrupt me and say I don't have to.

In summary:
 * Will factor out a log observer thingy.
 * Will rename `forceGarbageCollection` to `_forceGarbageCollection`.
 * Will _not_ change the constructor API for `TrialRunner`.
 * Will add test showing that passing `--force-gc` actually compels garbage collection.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml set owner to @exarkun|
|-|-|

I've made changes as described.

It's worth noting that I also discovered and fixed a recently introduced bug in the `TreeReporter`. The recent change made it so that the branch of the tree was determined by the id of the test, rather than its real class and module. There were two bugs in the new code: sometimes the id() method would return None (which is wrong) and the reporter would (correctly) explode; at other times, the id() would not contain enough segments.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @jml|
|-|-|

```ForceGarbageCollection``` class docstring has a typo in it.  ```simpleTest```, ```collect```, and ```makeRunner``` could use docstrings.

```test_forceGc``` docstring is a little bit unclear about whether the garbage collector is invoked before and each the entire suite or before and after each test method.  The test body doesn't clear this up either, since it makes those two cases identical.

```_ForceGarbageCollectionDecorator``` class docstring is also a bit unclear.  It'd be nice to say what the errors are added _to_, and to clarify that it's errors logged with t.p.log, not _raised_ errors.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @jml|
|-|-|

Looks good, go ahead and merge

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_13"></a>|@jml set status to `closed`|
|-|-|

(In [22231]) Make --force-gc apply even to tests made in a test_suite hook.

 * Author: jml
 * Reviewer: exarkun
 * Fixes [#2793](https://github.com/twisted/twisted/issues/2793)

Re-implements enforced garbage collection as a decorator, rather than a 
thing inside `TestCase`. This makes much of the code simpler.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
